compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InnerClassesLowering.kt

```diff
@@ -79,35 +79,37 @@ class InnerClassesLowering(val context: BackendContext) : ClassLoweringPass {
             override fun visitGetValue(expression: IrGetValue): IrExpression {
                 expression.transformChildrenVoid(this)
 
-                val implicitThisClass = expression.symbol.classForImplicitThis
-                if (implicitThisClass == null || implicitThisClass == irClass) return expression
+                val implicitThisClass = expression.symbol.classForImplicitThis ?: return expression
 
                 val startOffset = expression.startOffset
                 val endOffset = expression.endOffset
                 val origin = expression.origin
                 val function = currentFunction?.irElement as? IrFunction
-                val enclosingThisReceiver = function?.dispatchReceiverParameter ?: irClass.thisReceiver!!
-
-                var irThis: IrExpression = IrGetValueImpl(startOffset, endOffset, enclosingThisReceiver.symbol, origin)
-                var innerClass = irClass
-                while (innerClass != implicitThisClass) {
-                    if (!innerClass.isInner) {
-                        // Captured 'this' unrelated to inner classes nesting hierarchy, leave it as is -
-                        // should be transformed by closures conversion.
-                        return expression
-                    }
 
-                    irThis = if (function is IrConstructor && irClass == innerClass) {
-                        // Might be before a super() call (e.g. an argument to one), in which case the JVM bytecode verifier will reject
-                        // an attempt to access the field. Good thing we have a local variable as well.
-                        IrGetValueImpl(startOffset, endOffset, function.valueParameters[0].symbol, origin)
-                    } else {
-                        val outerThisField = context.declarationFactory.getOuterThisField(innerClass)
-                        IrGetFieldImpl(startOffset, endOffset, outerThisField.symbol, outerThisField.type, irThis, origin)
+                val receiverParameters = if (function == null || function is IrConstructor)
+                    listOfNotNull(irClass.thisReceiver)
+                else
+                    listOfNotNull(function.dispatchReceiverParameter, function.extensionReceiverParameter)
```

Comment:
`function.extensionReceiverParameter ` is redundant

Reply:
Actually as I see there is an `Outer.<anonymous>(): () -> String` passed to 'run' function. And during local declaration lowering this lambda is transformed to function with Outer extension receiver, that explicitly passed to it. So lower shouldn't try to lower any operation with this receiver.

Reply:
Actually, the extension receiver is of type `Inner` here, so `ok` *should* be lowered to `this$0.$outer.ok`. That's why this change fixes two tests; if the receiver was `Outer`, this change would have no effect as the first iteration of the loop would notice that and return the same IrGetValue.

Reply:
@pyos Next sample works incorrectly in JVM_IR:
```
class Outer {
    val z = 1;

    inner class Inner {
        fun Outer.foo() {
            z
        }
    }
}
```

Reply:
Seems that there are wrong  assumptions in `IrValueSymbol.classForImplicitThis` cause extension receiver also have '<this>' name

Reply:
Looks like you're right after all, the extension receiver in the lambda is `Outer`.